api/session.go

```diff
@@ -0,0 +1,372 @@
+		cert, err := cloud.Certificate()
+		if err != nil {
+			return err
+		}
+
+		joinIntent := types.SessionJoinPost{
+			Version:     cloudMDNS.Version,
+			Name:        state.Name(),
+			Address:     session.Address,
+			Certificate: string(cert.PublicKey()),
+			Services:    session.Services,
+		}
+
+		argonFormat, err := trust.NewDefaultHMACFormatArgon()
+		if err != nil {
+			return err
+		}
+
+		hmac := trust.NewHMAC(trust.NewDefaultHMACConf(sessionPassphrase, HMACMicroCloud10)).WithFormat(argonFormat)
+		header, err := hmac.AuthorizationHeader(joinIntent)
+		if err != nil {
+			return err
+		}
+
+		// Confirm join intent.
+		// This request uses polling to wait for confirmation from the other side.
+		g.Go(func() error {
+			conf := cloudClient.AuthConfig{
+				HMAC: header,
+				// We already know the certificate of the joiner for TLS verification.
+				TLSServerCertificate: remoteCert,
+			}
+
+			return cloud.RequestJoinIntent(ctx, intent.Address, conf, joinIntent)
+		})
+	}
+
+	err = g.Wait()
+	if err != nil {
+		return fmt.Errorf("Failed to confirm join intents: %w", err)
+	}
+
+	err = gw.Write(types.Session{
+		Accepted: true,
+	})
+	if err != nil {
+		return fmt.Errorf("Failed to send confirmation: %w", err)
+	}
+
+	return nil
+}
+
+func handleJoiningSession(state *state.State, sh *service.Handler, gw *cloudClient.WebsocketGateway) error {
+	session := types.Session{}
+	err := gw.ReceiveWithContext(gw.Context(), &session)
+	if err != nil {
+		return fmt.Errorf("Failed to read session start message: %w", err)
+	}
+
+	err = sh.StartSession(session.Passphrase)
+	if err != nil {
+		return fmt.Errorf("Failed to start session: %w", err)
+	}
+
+	defer func() {
+		err := sh.StopSession()
+		if err != nil {
+			logger.Error("Failed to stop session", logger.Ctx{"err": err})
+		}
+	}()
+
+	iface, err := net.InterfaceByName(session.Interface)
+	if err != nil {
+		return fmt.Errorf("Failed to lookup interface by name: %w", err)
+	}
+
+	// No address selected, try to lookup system.
+	if session.TargetAddress == "" {
+		lookupCtx, cancel := context.WithTimeout(gw.Context(), session.LookupTimeout)
+		defer cancel()
+
+		peer, err := cloudMDNS.LookupPeer(lookupCtx, iface, cloudMDNS.Version)
+		if err != nil {
+			return err
+		}
+
+		session.TargetAddress = peer.Address
+	}
+
+	// Get the remotes fingerprint from the used certificate.
+	url := api.NewURL().Scheme("https").Host(fmt.Sprintf("%s:%d", session.TargetAddress, service.CloudPort))
+	peerCert, err := shared.GetRemoteCertificate(url.String(), "MicroCloud")
+	if err != nil {
+		return fmt.Errorf("Failed to retrieve remote certificate: %w", err)
+	}
+
+	session.TargetFingerprint = shared.CertFingerprint(peerCert)
+
+	// Get the remotes name.
+	cloud := sh.Services[types.MicroCloud].(*service.CloudService)
+	peerStatus, err := cloud.RemoteStatus(gw.Context(), peerCert, session.TargetAddress)
+	if err != nil {
+		return fmt.Errorf("Failed to retrieve cluster status: %w", err)
+	}
```

Comment:
Would it be worth it to make this a bit more efficient by returning all the information we need from `RequestJoinIntent`? It would reduce the network back/forth by 1, but it would also push the client notification back a bit.

Reply:
You mean the extra call for getting the cert? 

I was deciding for this in favor of drilling down the microcluster query methods in order to grep something from the HTTP's response `TLS` field.

Do you think it would be worth exposing this in another microcluster query func or should we use the `http` request tooling instead of the microcluster query func inside `RequestJoinIntent`?

Reply:
No I meant the additional query to `/1.0` to fetch the cluster name. I wonder if `RequestJoinIntent` should return a response body that contains it so we don't need the additional request over the network.

Reply:
Fixed by getting the fingerprint from the actual TLS connection when `RequestJoinIntent` is sent.
The name is fetched from the already existing microcluster status endpoint.